tests/functional/gin.cpp

```diff
@@ -0,0 +1,268 @@
+			if (i < rank) {
+				++local_rank;
+			}
+		}
+	}
+
+	/* Set CUDA device for subsequent device memory allocation, in case GDR is used */
+	NCCL_OFI_TRACE(NCCL_NET, "Using CUDA device %d for memory allocation", local_rank);
+
+	/* Get external Network from NCCL-OFI library */
+	(void)get_extNet; /* unused warning */
+	get_ext(&extNet, &extGin);
+	if (extNet == nullptr || extGin == NULL) {
+		res = ncclInternalError;
+		return res;
+	}
+
+	OFINCCLCHECK(extNet->init(&logger, nullptr));
+
+	void *ginCtx = nullptr;
+
+	/* Init API */
+	OFINCCLCHECK(extGin->init(&ginCtx, 0, &logger));
+	NCCL_OFI_INFO(NCCL_NET, "Process rank %d started. NCCLNet device used on %s is %s.",
+		      rank, &all_proc_name[PROC_NAME_IDX(rank)], extGin->name);
+
+	/* Devices API */
+	OFINCCLCHECK(extGin->devices(&ndev));
+	NCCL_OFI_INFO(NCCL_NET, "Received %d network devices", ndev);
+
+	/* Indicates if NICs support GPUDirect */
+	std::vector<int> test_support_gdr(ndev);
+
+	/* Get Properties for the device */
+	for (dev = 0; dev < ndev; dev++) {
+		ncclNetProperties_v11_t props = {};
+		OFINCCLCHECK(extGin->getProperties(dev, &props));
+
+		/* Set CUDA support */
+		test_support_gdr[dev] = is_gdr_supported_nic(props.ptrSupport);
+	}
+
+	dev = local_rank % ndev;
+
+	NCCL_OFI_TRACE(NCCL_INIT, "Rank %d uses %d device for communication", rank, dev);
+
+	if (test_support_gdr[dev] == 1) {
+		NCCL_OFI_INFO(NCCL_INIT | NCCL_NET,
+				"Network supports communication using CUDA buffers. Dev: %d", dev);
+		buffer_type = NCCL_PTR_CUDA;
+	} else {
+		/* We aren't currently interested in the non-GDR use case for GIN */
+		NCCL_OFI_WARN("Network does not support communication using CUDA buffers. Dev: %d", dev);
+		return 1;
+	}
+
+	void *listenComm = nullptr;
+	OFINCCLCHECK(extGin->listen(ginCtx, dev, handles[rank], &listenComm));
+	assert(listenComm);
+
+	/* Gather handles from all ranks */
+	MPI_Allgather(MPI_IN_PLACE, 0, MPI_DATATYPE_NULL, handles.data(), NCCL_NET_HANDLE_MAXSIZE, MPI_CHAR, MPI_COMM_WORLD);
+
+	/* Prepare handles void** array */
+	for (int i = 0; i < nranks; ++i) {
+		handles_ptrs[i] = &(handles[i]);
+	}
+
+	void *collComm = nullptr;
+	OFINCCLCHECK(extGin->connect(ginCtx, handles_ptrs.data(), nranks, rank, listenComm,
+				     &collComm));
+	assert(collComm != nullptr);
+
+	void *buff = nullptr;
+	OFINCCLCHECK(allocate_buff(&buff, SEND_SIZE, buffer_type));
+	OFINCCLCHECK(initialize_buff(buff, SEND_SIZE, buffer_type));
+
+	/* Collective memory registration */
+	void *mhandle = nullptr, *gin_mhandle = nullptr;
+	constexpr uint64_t mrFlags = 0; /* TODO FORCE_SO */
+	OFINCCLCHECK(extGin->regMrSym(collComm, buff, SEND_SIZE, buffer_type, mrFlags,
+				      &mhandle, &gin_mhandle));
+	assert(mhandle != nullptr && gin_mhandle != nullptr);
+
+	void *signal_buf = nullptr;
+	OFINCCLCHECK(allocate_buff(&signal_buf, sizeof(uint64_t), buffer_type));
+	OFINCCLCHECK(initialize_buff(buff, SEND_SIZE, buffer_type, 0));
+	CUDACHECK(cudaStreamSynchronize(cudaStreamDefault));
```

Comment:
If the `cudaStreamSynchronize` is called to ensure a `cudaMemset` in `initialize_buff` has completed, why not have it inside `initialize_buff`?